_posts/2017-10-11-cross-site-request-forgery.markdown

```diff
@@ -0,0 +1,217 @@
+
+It is quite common to leave REST APIs accessible to the end-users, either because they are being used by JavaScript based user interfaces, or to expose the application's resources to external clients. The main difference between the REST APIs and regular HTML pages is that browsers tend to retrieve the pages before trying to submit them, hence it is easy to make the CSRF token available to them. When it comes to REST APIs, quite often this is not the case, the API calls are most of the time one-off requests, which makes this approach rather difficult:
+
+* If the CSRF token is the same or a derived/computed value from the session ID, the problem is that the session ID optimally should not be available for JavaScript applications -- for example because the cookie is [HttpOnly].
+* If the CSRF token is an encrypted token, then there needs to be a way for the client to retrieve this value before actually using it. This however requires a reasonable mechanism to distribute the encrypted token (as part of an authentication response for example), but one cannot always guarantee that the end-user have used the application in question to authenticate. Forcing API users to send two requests per operation (one for getting a CSRF token, one for actually performing the operation) is probably not going to scale well.
+
+An alternative solution to CSRF tokens is the usage of custom headers, however those have their pros and cons as well:
+
+* Requiring a custom header on the REST endpoint deals with simple browser requests (think of the first and second CSRF examples), but requires [additional CORS setup][Access-Control-Allow-Headers] when working with AJAX clients.
+* If an origin is whitelisted in CORS to make REST calls, then the application is only protected from CSRF attacks as long as the trusted origin is not vulnerable to Content Injection attacks.
+
+I should probably also highlight that restricting which HTTP verbs can be used by the REST clients is not an adequate protection either: various [frameworks][symfony method override] or [containers][nodejs method override] even allow the modification of the request method based on request parameter/header.
+
+So all in all, when it comes to REST APIs, implementing CSRF protections is not really that straightforward and requires careful design process. Combining one of the above approaches with Origin/Referer header validation can provide a decent level of protection against CSRF attacks at REST APIs, however I've got to say that sounds like a lot of work...
+
+### Future of CSRF attacks
+
+Fortunately CSRF doesn't have a long term future, the currently draft spec of [SameSite] cookies will prevent these attacks completely. The new SameSite flag on cookies instructs the browser to only include the cookie values when the top level navigation origin matches the cookie's domain. The support for SameSite cookies is [work in progress][caniuse], but the future looks bright at least.
```

Comment:
I would say that the problem will probably move elsewhere. Rather than trying to trick the browser into making requests with the user's cookies, the attacks will instead try to trick ever-more-complex Javascript into doing this. For instance, single-page apps that perform XHR requests depending on what is in the URL fragment are often vulnerable in this way: `<img src="https://bank.example.com/some/spa#transfer&from=...&to=...&amount"/>`

Reply:
Good point, I shouldn't be that cheerful :)